### PR TITLE
Handle duplicated registry ids in 1.20.3->.5

### PR DESCRIPTION
**Pull request by FlorianMichael**
Minecraft 1.20.4 and earlier silently overrides the previous values if multiple registry elements have the same id (Tested using a Fabric 1.20.4 mod by parsing a registry sent by a server and then printing the parsed objects).

### CODE REVIEW

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java [hunk 1]
```diff
@@ -112,17 +114,23 @@ public void registerPackets() {
                 final CompoundTag entryTag = (CompoundTag) entry.getValue();
                 final String type = entryTag.getString("type");
                 final ListTag<CompoundTag> valueTag = entryTag.getListTag("value", CompoundTag.class);
-                RegistryEntry[] registryEntries = new RegistryEntry[valueTag.size()];
+
+                // Calculate number of entries (exclude duplicated ids)
+                RegistryEntry[] registryEntries = new RegistryEntry[valueTag.stream().map(e -> e.getInt("id")).distinct().toArray().length];
                 boolean requiresDummyValues = false;
                 int entriesLength = registryEntries.length;
+                Set<Integer> ids = new HashSet<>();
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java:122]: IntSet

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java [hunk 2]
```diff
@@ -112,17 +114,23 @@ public void registerPackets() {
...
                 for (final CompoundTag tag : valueTag) {
                     final String name = tag.getString("name");
                     final int id = tag.getInt("id");
-                    entriesLength = Math.max(entriesLength, id + 1);
-                    if (id >= registryEntries.length) {
-                        // It was previously possible to have arbitrary ids
-                        registryEntries = Arrays.copyOf(registryEntries, Math.max(registryEntries.length * 2, id + 1));
-                        requiresDummyValues = true;
+                    if (!ids.contains(id)) { // Override duplicated id without incrementing entries length
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java:126]: if (!ids.add(id))

**Review by kennytv — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java (hunk 2 above)
Comment (FlorianMichael) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java]: fixed

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java (hunk 1 above)
Comment (FlorianMichael) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java]: fixed

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/data/DamageTypes1_20_3.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        return create(messageId, "when_caused_by_living_non_player", exhaustion);
+    }
+
+    private static CompoundTag create(String messageId, String scaling, float exhaustion, String damageEffects, String deathMessageType) {
+        final CompoundTag tag = new CompoundTag();
+        tag.putString("message_id", messageId);
+        tag.putString("scaling", scaling);
+        tag.putFloat("exhaustion", exhaustion);
+        if (damageEffects != null) tag.putString("effects", damageEffects);
+        if (deathMessageType != null) tag.putString("death_message_type", deathMessageType);
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/data/DamageTypes1_20_3.java:102]: new line and braces

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java [hunk 4]
```diff
@@ -135,15 +138,25 @@ public void registerPackets() {
...
                     final int length = registryEntries.length;
                     registryEntries = Arrays.copyOf(registryEntries, length + 1);
                     final CompoundTag spitData = new CompoundTag();
                     spitData.putString("scaling", "when_caused_by_living_non_player");
                     spitData.putString("message_id", "mob");
                     spitData.putFloat("exhaustion", 0.1F);
                     registryEntries[length] = new RegistryEntry("minecraft:spit", spitData);
+
+                    // Fill in missing damage types with 1.20.3/4 defaults
+                    final Set<String> registryEntryKeys = Arrays.stream(registryEntries).map(RegistryEntry::key).collect(Collectors.toSet());
```
Comment (kennytv) [inline common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java:152]: In the map call you need to add Key.stripmcnamespace

**Review by kennytv — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/data/DamageTypes1_20_3.java (hunk 3 above)
Comment (FlorianMichael) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/data/DamageTypes1_20_3.java]: Fixed

File: common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java (hunk 4 above)
Comment (FlorianMichael) [reply common/src/main/java/com/viaversion/viaversion/protocols/protocol1_20_5to1_20_3/rewriter/EntityPacketRewriter1_20_5.java]: Fixed

**Review by kennytv — APPROVED** [review]

**Pull request merged**